src/v/cloud_topics/level_one/common/file_arena.cc

```diff
@@ -0,0 +1,310 @@
+/*
+ * Copyright 2025 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "cloud_topics/level_one/common/file_arena.h"
+
+#include "cloud_topics/level_one/common/file_arena_probe.h"
+#include "ssx/future-util.h"
+#include "ssx/semaphore.h"
+#include "utils/human.h"
+
+#include <chrono>
+
+using namespace std::chrono_literals;
+
+namespace cloud_topics::l1 {
+
+staging_file_with_reservation::staging_file_with_reservation(
+  std::unique_ptr<staging_file> file, ssx::semaphore_units disk_reservation)
+  : _file(std::move(file))
+  , _disk_reservation(std::move(disk_reservation)) {}
+
+ss::future<size_t> staging_file_with_reservation::size() {
+    return _file->size();
+}
+
+ss::future<ss::output_stream<char>>
+staging_file_with_reservation::output_stream() {
+    return _file->output_stream();
+}
+
+ss::future<> staging_file_with_reservation::finalize() {
+    size_t units = _disk_reservation.count();
+    size_t file_size = co_await _file->size();
```

Comment:
The callers know the size here, they could pass in the size and save the syscall. Maybe over optimization?

Reply:
I would just say make it required and the user can make the syscall